src/turbolinks/location.coffee

```diff
@@ -6,17 +6,16 @@ class Turbolinks.Location
       new this value
 
   constructor: (url = "") ->
-    linkWithAnchor = document.createElement("a")
-    linkWithAnchor.href = url.toString()
+    link = document.createElement("a")
+    link.href = url.toString()
 
-    @absoluteURL = linkWithAnchor.href
+    @absoluteURL = link.href
+    @anchor = @absoluteURL.match(/#(.*)$/)?[1] ? null
 
-    anchorLength = linkWithAnchor.hash.length
-    if anchorLength < 2
-      @requestURL = @absoluteURL
+    @requestURL = if @anchor?
+      @absoluteURL.slice(0, -(@anchor.length + 1))
     else
-      @requestURL = @absoluteURL.slice(0, -anchorLength)
-      @anchor = linkWithAnchor.hash.slice(1)
+      @absoluteURL
```

Comment:
I'm having a hard time digesting the changes here. Can you summarize them? Why switch from `link.hash` to a regex?

Reply:
This change is a workaround for the (inconsistent) native behaviour of `link.hash`. The native behaviour is as follows:

```js
link.href = '#foo'
link.hash // '#foo'

link.href = '#'
link.hash // ''

link.href = 'http://example.com'
link.hash // ''
```

This makes it impossible to distinguish whether a location has an empty hash or no hash at all.

To fix this, the `anchor` is just the text part of the hash or `null` if the URL does not contain a hash at all:

```js
link.href = '#foo'
@anchor // 'foo'

link.href = '#'
@anchor // ''

link.href = 'http://example.com'
@anchor // null
```

This is required for the [`locationHasSamePageAnchor`](https://github.com/turbolinks/turbolinks/pull/285/files#diff-511f2c4262a37015ef9a6d1a13d4a36dR163) method which calls `@location.anchor?`. Without this change, `@location.anchor?` will always be `true`.

In terms of the outcome, this change prevents Turbolinks from making requests when navigating to '#' (https://github.com/turbolinks/turbolinks/issues/192).

Let me know if you'd like me to add comments to the code for this.

Reply:
Hey @javan, do you have any thoughts on this approach?

Reply:
Cool, thanks for the explanation! 

Splitting the constructor code out into intention revealing methods/helpers would help clarify. Your changes don't make the code any more confusing so I'm not suggesting you should change anything. Just trying to grok the before / after.

Reply:
Yes, I’d happily tidy this up at some point; perhaps at the same time as refactoring the methods to use more `<a>` properties (as mentioned here: https://github.com/turbolinks/turbolinks/pull/270#issuecomment-307450179). Let me know if there’s anything else this PR needs. Thanks!

Reply:
I think we could tidy this up with:

```coffeescript
[@requestURL, @anchor] = @absoluteURL.split('#')
```